ProcessLib/UncoupledProcessesTimeLoop.cpp

```diff
@@ -405,35 +458,90 @@ bool solveOneTimeStepOneProcess(GlobalVector& x, std::size_t const timestep,
         output_control.doOutputNonlinearIteration(
             process, process_data.process_output, timestep, t, x, iteration);
     };
 
     bool nonlinear_solver_succeeded =
-        nonlinear_solver.solve(x, post_iteration_callback);
+        nonlinear_solver.solve(x, coupled_term, post_iteration_callback);
 
     auto& mat_strg = *process_data.mat_strg;
     time_disc.pushState(t, x, mat_strg);
 
-    process.postTimestep(x);
-
     return nonlinear_solver_succeeded;
 }
 
 UncoupledProcessesTimeLoop::UncoupledProcessesTimeLoop(
     std::unique_ptr<NumLib::ITimeStepAlgorithm>&& timestepper,
     std::unique_ptr<Output>&& output,
-    std::vector<std::unique_ptr<SingleProcessData>>&& per_process_data)
+    std::vector<std::unique_ptr<SingleProcessData>>&& per_process_data,
+    const unsigned global_coupling_max_iterations,
+    std::unique_ptr<NumLib::ConvergenceCriterion>&& glb_coupling_conv_crit)
     : _timestepper{std::move(timestepper)},
       _output(std::move(output)),
-      _per_process_data(std::move(per_process_data))
+      _per_process_data(std::move(per_process_data)),
+      _global_coupling_max_iterations(global_coupling_max_iterations),
+      _global_coupling_conv_crit(std::move(glb_coupling_conv_crit))
 {
 }
 
+bool UncoupledProcessesTimeLoop::setCoupledSolutions()
+{
+    if (!_global_coupling_conv_crit && _global_coupling_max_iterations == 1)
+        return false;
+
+    unsigned pcs_idx = 0;
+    _solutions_of_coupled_processes.resize(_per_process_data.size());
+    for (auto& spd : _per_process_data)
+    {
+        BaseLib::RunTime time_timestep_process;
+        time_timestep_process.start();
+
+        auto const& coupled_processes = spd->coupled_processes;
+        std::map<ProcessType, GlobalVector const*> coupled_xs;
+        auto it = coupled_processes.begin();
+        while (it != coupled_processes.end())
+        {
+            ProcessLib::Process const& coupled_process = it->second;
+            auto const found_item = std::find_if(
+                _per_process_data.begin(),
+                _per_process_data.end(),
+                [&coupled_process](
+                    std::unique_ptr<SingleProcessData> const& item) {
+                    return coupled_process.getProcessType() ==
+                           item->process.getProcessType();
+                });
+
+            if (found_item != _per_process_data.end())
+            {
+                // Id of the coupled process:
+                const std::size_t c_id =
+                    found_item - _per_process_data.begin();
+
+                BaseLib::insertMapIfKeyUniqueElseError(coupled_xs, it->first,
+                                                       _process_solutions[c_id],
+                                                       "global_coupled_x");
+            }
+            it++;
+        }
+        _solutions_of_coupled_processes[pcs_idx] = coupled_xs;
+
+        const auto x = _process_solutions[pcs_idx];
```

Comment:
Changed to ``const auto x = _process_solutions[pcs_idx];``. With ``auto const* x`` and ``NumLib::GlobalVectorProvider::provider.getVector(x)``, ``getVector(std::size_t& id)`` is called instead of ``getVector(GlobalVector const& x)``.